### PR TITLE
executor: fix the forever hang in index merge join when oom occurs

### LINKED ISSUE #17394: query stuck in index lookup merge join
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Mock oom panic in `innerMergeWorker.handleTask`:

image

2. Execute below SQL:

```sql
create table t1 (a int,b int,index idx(a));
create table t2 (a int,b int,index idx(a));
insert into t1 values (1,1),(2,2),(3,3),(2000,2000)；
insert into t2 values (1,1),(2,2),(3,3),(2000,2000)；
select /*+INL_MERGE_JOIN(t1,t2) */ * from t1,t2 where t1.a=t2.a; -- query will get stuck.
```

### 2. What did you expect to see? (Required)

Query not stuck, just return error.

### 4. Affected version (Required)

```sql
mysql>select tidb_version();
+-------------------------------------------------------------------+
| tidb_version()                                                    |
+-------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-502-g43965127c-dirty               |
| Edition: Community                                                |
| Git Commit Hash: 43965127c69f7ecb1821e622ecfdaa3de4f334f1         |
| Git Branch: master                                                |
| UTC Build Time: 2020-05-25 02:21:08                               |
| GoVersion: go1.13                                                 |
| Race Enabled: false                                               |
| TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306 |
| Check Table Before Drop: false                                    |
+-------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close #17394 
Problem Summary: As the title says.

### What is changed and how it works?

What's Changed: close the results channel for the panic inner task.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- fix the forever hang in index merge join when oom occurs

### CODE REVIEW

File: executor/index_lookup_merge_join.go [hunk 1]
```diff
@@ -387,13 +390,17 @@ func (omw *outerMergeWorker) increaseBatchSize() {
 func (imw *innerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup, cancelFunc context.CancelFunc) {
```
Comment (zz-jason) [inline executor/index_lookup_merge_join.go:390]: It's better to use `util.WithRecovery()` to start this goroutine, thus you don't need to copy the code to print the stack.

You can log a message in the `recoverFn` to indicate that this worker is an inner worker of index merge join.

Comment (lzmhhh123) [reply executor/index_lookup_merge_join.go:390]: `util.WithRecovery()` execute in the function which starts the goroutine. But in this case, we need to `close(task.results)`, which is defined in inner goroutine function. `util.WithRecovery()` can't do this.

File: executor/index_lookup_merge_join_test.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	. "github.com/pingcap/check"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/util/testkit"
+)
+
+func (s *testSuite9) TestIndexLookupMergeJoinHang(c *C) {
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/IndexMergeJoinMockOOM", `return(true)`), IsNil)
+	defer c.Assert(failpoint.Disable("github.com/pingcap/tidb/executor/IndexMergeJoinMockOOM"), IsNil)
```
Comment (XuHuaiyu) [inline executor/index_lookup_merge_join_test.go:11]: defer func(){}()

File: executor/index_lookup_merge_join_test.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("drop table if exists t1, t2")
+	tk.MustExec("create table t1 (a int,b int,index idx(a))")
+	tk.MustExec("create table t2 (a int,b int,index idx(a))")
+	tk.MustExec("insert into t1 values (1,1),(2,2),(3,3),(2000,2000)")
+	tk.MustExec("insert into t2 values (1,1),(2,2),(3,3),(2000,2000)")
+	// Do not hang in index merge join when OOM occurs.
+	tk.MustExec("select /*+ INL_MERGE_JOIN(t1, t2) */ * from t1, t2 where t1.a = t2.a")
```
Comment (XuHuaiyu) [inline executor/index_lookup_merge_join_test.go:20]: MustQuery

**Pull request merged**